### PR TITLE
ARROW-10887: [Doc][C++] Document C++ IPC API

### PR DESCRIPTION
**Pull request by pitrou**
Also add some tables of contents here and there for long nested documents.

### CODE REVIEW

File: docs/source/cpp/ipc.rst [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+.. seealso::
+   :ref:`Arrow IPC format specification <format-ipc>`.
+
+   :doc:`API reference for IPC readers and writers <api/ipc>`.
+
+Arrow C++ provides abstracted readers and writers for the Arrow IPC format.
+Actual input/output is handled through the :doc:`IO interfaces <io>`.
+For reading, there is also an event-driven API that allows to feed arbitrary
+data into the IPC decoding layer.
```
Comment (bkietz) [inline docs/source/cpp/ipc.rst:34]:
```suggestion
Arrow C++ provides readers and writers for the Arrow IPC format which wrap
lower level input/output, handled through the :doc:`IO interfaces <io>`.
For reading, there is also an event-driven API that enables feeding
arbitrary data into the IPC decoding layer asynchronously.
```

File: docs/source/cpp/api/ipc.rst [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+----------
+
+.. doxygenstruct:: arrow::ipc::ReadStats
+   :members:
+
+Writing IPC streams and files
+=============================
+
+Blocking API
+------------
```
Comment (bkietz) [inline docs/source/cpp/api/ipc.rst:77]:
```suggestion
------------

The IPC stream format is only optionally terminated, whereas the IPC file format
must include a terminating footer. Thus a writer of the IPC file format must be
explicitly finalized with :func:`arrow::RecordBatchWriter::Close()` or the resulting
file will be corrupt.
```

**Review by bkietz — CHANGES_REQUESTED** [review]
Just a few notes

File: docs/source/cpp/api/ipc.rst (hunk 2 above)
Comment (pitrou) [reply docs/source/cpp/api/ipc.rst:77]: Neat suggestion, thank you.

**Pull request closed without merging**